compose/service.py

```diff
@@ -770,6 +765,47 @@ def parse_volume_spec(volume_config):
     return VolumeSpec(external, internal, mode)
 
 
+def build_volume_from(volume_from_spec):
+    volumes_from = []
+    if isinstance(volume_from_spec.source, Service):
+        containers = volume_from_spec.source.containers(stopped=True)
+        if not containers:
+            volumes_from = ["{}:{}".format(volume_from_spec.source.create_container().id, volume_from_spec.mode)]
+        else:
+            volumes_from = ["{}:{}".format(container.id, volume_from_spec.mode) for container in containers]
+    elif isinstance(volume_from_spec.source, Container):
+        volumes_from = ["{}:{}".format(volume_from_spec.source.id, volume_from_spec.mode)]
+    return volumes_from
+
+
+def parse_volume_from_spec(volume_from_config, get_service, client):
```

Comment:
Sorry, I was unclear. I was thinking that just the string splitting and concatenation logic should be moved into utility functions - the part where we determine whether the source is a service or a container (as well as the corresponding logic in `Service._get_volumes_from` where we check whether we're looking at an instance of Service or Container) can stay where it is.